crypto: add support for parsing SSE-C/SSE-S3 metadata

aead opened the pull request:
## Description
This commit adds support for detecting and parsing
SSE-C/SSE-S3 object metadata. With the `IsEncrypted`
functions it is possible to determine whether an object
seems to be encrypted. With the `ParseMetadata` functions
it is possible to validate such metadata and extract the
SSE-C/SSE-S3 related values.

It also fixes some naming issues.

## Motivation and Context
SSE-C, SSE-S3, KMS

## How Has This Been Tested?
unit tests: `go test -v github.com/minio/minio/cmd/crypto`

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

Commit:
crypto: add support for parsing SSE-C/SSE-S3 metadata

This commit adds support for detecting and parsing
SSE-C/SSE-S3 object metadata. With the `IsEncrypted`
functions it is possible to determine whether an object
seems to be encrypted. With the `ParseMetadata` functions
it is possible to validate such metadata and extract the
SSE-C/SSE-S3 related values.

It also fixes some naming issues.

poornas reviewed (dismissed):
LGTM

harshavardhana commented on cmd/crypto/metadata_test.go:
```diff
@@ -0,0 +1,259 @@
+package crypto
```
Adding missing license headers

aead replied:
Oh forgot, fixed.

harshavardhana commented on cmd/crypto/metadata.go:
```diff
@@ -0,0 +1,149 @@
+package crypto
```
Add missing license header..

aead replied:
Done

Commit:
crypto: add functions for creating SSE object metadata

This commit adds functions for creating SSE-S3 and
SSE-C metadata. It also adds a `CreateMultipartMetadata`
for creating multipart metadata.

For all functions unit tests are included.

donatello commented on cmd/crypto/metadata.go:
```diff
@@ -0,0 +1,213 @@
+	}
+	b64SealedKey, ok := metadata[S3SealedKey]
+	if !ok {
+		return keyID, kmsKey, sealedKey, Error{"The object metadata is missing the internal sealed key for SSE-S3"}
```
Why are some errors being named (in `crypto/error.go`) and some others not?

aead replied:
Two errors (e.g. `errMissingIV` and `errMissingSealAlgorithm`) can happen for SSE-C and SSE-S3. (Those metadata entries are shared). Other errors like `errMissingKMSKeyID` can only happen for SSE-S3, not for SSE-C. So `errMissingIV` (for example) is a named / bound to identifier to avoid repeating `Error{"The object metadata is missing the IV entry"}`. Non-named errors can only happen at the places they are used.

So basically avoid repeating code. FYI - The error msg will be sent to client (as `XMinioObjectTampered` API error code). Avoiding multiple copies of the same error string helps to ensure that the client get's an accurate error message (in contrast to AWS which returns different error msgs for equal error conditions in different scenarios - different error msgs refer to the same root cause)

poornas reviewed (approved):
LGTM
